x/programs/cmd/simulator/src/lib.rs

```diff
@@ -227,105 +228,95 @@ impl Client {
+        let child = Command::new(path)
+            .arg("interpreter")
+            .arg("--cleanup")
+            .arg("--log-level")
+            .arg("error")
+            .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            .spawn()?;
+
+        let (Some(stdin), Some(stdout)) = (child.stdin, child.stdout) else {
+            panic!("could not attach to the child process stdio");
+        };
+        let lines = BufReader::new(stdout).lines();
+
+        let mut sim = Self { stdin, lines };
+
+        let _ = sim.read_response()?; // intepreter ready
+
+        Ok(sim)
     }
 
-    for line in String::from_utf8(output.stdout)?
-        .lines()
-        .filter(|line| !line.trim().is_empty())
-    {
-        let item = serde_json::from_str(line)
-            .map_err(|e| format!("failed to parse output to json: {e}"))?;
-        items.push(item);
+    fn run_step(
+        &mut self,
+        caller_key: &String,
+        step: &Step,
+    ) -> Result<PlanResponse, Box<dyn Error>> {
+        let run_command = "run --stdin\n";
+        self.write_stdin(run_command.as_bytes())?;
+        let _ = self.read_response()?;
+
+        let step = SimulatorStep { caller_key, step };
+        let mut input =
+            serde_json::to_string(&step).map_err(|e| format!("failed to serialize json: {e}"))?;
+        input.push('\n');
+        self.write_stdin(input.as_bytes())?;
+        let resp = self.read_response()?;
+
+        Ok(resp)
     }
 
-    Ok(items)
-}
+    fn read_response(&mut self) -> Result<PlanResponse, Box<dyn Error>> {
+        let text: String = self
+            .lines
+            .next()
+            .ok_or("EOF")?
+            .map_err(|e| format!("failed to read from stdout: {e}"))?;
 
-fn run_step<P, T>(path: P, plan: &Plan) -> Result<T, Box<dyn Error>>
-where
-    T: serde::de::DeserializeOwned + serde::Serialize,
-    P: AsRef<OsStr>,
-{
-    let output = cmd_output(path, plan)?;
-
-    if !output.status.success() {
-        println!("stderr");
-        for line in String::from_utf8_lossy(&output.stderr).lines() {
-            println!("{line}");
-        }
-        println!("stdout");
-        for line in String::from_utf8_lossy(&output.stdout).lines() {
-            println!("{line}");
-        }
-        return Err(String::from_utf8(output.stdout)?.into());
+        let resp = serde_json::from_str(&text)
+            .map_err(|e| format!("failed to parse output to json: {e}"))?;
+
+        Ok(resp)
     }
 
-    let resp: T = serde_json::from_str(String::from_utf8(output.stdout)?.as_ref())
-        .map_err(|e| format!("failed to parse output to json: {e}"))?;
+    fn write_stdin(&mut self, bytes: &[u8]) -> Result<(), String> {
+        self.stdin
+            .write_all(bytes)
+            .map_err(|e| format!("failed to write to stdin: {e}"))?;
+        self.stdin
+            .flush()
+            .map_err(|e| format!("failed to flush from stdin: {e}"))?;
 
-    Ok(resp)
+        Ok(())
+    }
```

Comment:
With this, you're forced to flush every time you write. It's better to just deal with something that implements the `Write` trait directly so that you can call flush when you want.